src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,349 @@
+/-
+Copyright (c) 2022 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+
+import geometry.manifold.mfderiv
+import analysis.complex.upper_half_plane.functions_bounded_at_infty
+import analysis.complex.upper_half_plane.topology
+import number_theory.modular_forms.slash_invariant_forms
+/-!
+# Modular forms
+
+This file defines modular forms and proves some basic properties about them.
+
+We begin by defining modular forms and cusp forms as extension of `slash_invariant_forms` then we
+define the space of modular forms, cusp forms and prove that the product of two modular forms is a
+modular form.
+-/
+
+open complex upper_half_plane
+
+open_locale topological_space manifold upper_half_plane
+
+noncomputable theory
+
+instance upper_half_plane.charted_space : charted_space ℂ ℍ :=
+upper_half_plane.open_embedding_coe.singleton_charted_space
+
+instance upper_half_plane.smooth_manifold_with_corners : smooth_manifold_with_corners 𝓘(ℂ) ℍ :=
+open_embedding.singleton_smooth_manifold_with_corners 𝓘(ℂ) (upper_half_plane.open_embedding_coe)
+
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) _) _
+
+local notation `GL(` n `, ` R `)`⁺ := matrix.GL_pos (fin n) R
+
+local notation `SL(` n `, ` R `)` := matrix.special_linear_group (fin n) R
+
+section modular_form
+
+open modular_form
+
+variables (F : Type*) (Γ : subgroup SL(2, ℤ)) (k : ℤ)
+
+local notation f `∣[`:73 k:0, A `]` :72 := slash_action.map ℂ k A f
+
+set_option old_structure_cmd true
+
+/--These are `slash_invariant_form`'s that are holomophic and bounded at infinity. -/
+structure modular_form extends slash_invariant_form Γ k :=
+(hol' : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (to_fun : ℍ → ℂ))
+(bdd_at_infty' : ∀ (A : SL(2, ℤ)), is_bounded_at_im_infty (to_fun ∣[k, A]))
+
+/-- The `slash_invariant_form` associated to a `modular_form`. -/
+add_decl_doc modular_form.to_slash_invariant_form
+
+/--These are `slash_invariant_form`s that are holomophic and zero at infinity. -/
+structure cusp_form extends slash_invariant_form Γ k :=
+(hol' : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (to_fun : ℍ → ℂ))
+(zero_at_infty' : ∀ (A : SL(2, ℤ)), is_zero_at_im_infty (to_fun ∣[k, A]))
```

Comment:
Why not define cusp forms explicitly as a submodule of modular forms? I think it would save you some work at the end of the file.

Reply:
Well I was just following @j-loreaux suggestion here: https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2313250.20Modular.20form.20definition/near/303533157

Reply:
@hrmacbeth Do you have an opinion on this?

Reply:
I don't know the `fun_like` philosophy very well, I would defer to @mcdoll and @j-loreaux who suggested this design.

Reply:
My impression was that by having these classes then some lemmas would be automatic for the extensions. If I were to define them as submodules would this still work?